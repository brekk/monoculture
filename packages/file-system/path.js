import { join, normalize } from 'node:path'
import { curry } from 'ramda'

// pathJoin :: String -> String -> String
export const pathJoin = curry((pwd, x) => {
  // TODO: this should ostensibly be the province of a type system to ensure this safety
  if (typeof pwd !== 'string' || typeof x !== 'string') {
    throw new Error('Cannot normalize bad paths.')
  }
  return join(pwd, normalize(x))
})
