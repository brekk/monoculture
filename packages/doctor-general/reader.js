import { log } from './log'
import { chain, when, curry, flatten, map, pipe, propOr } from 'ramda'
import { parallel } from 'fluture'
import { signal } from 'kiddo'
import { readJSONFile, readDirWithConfig } from 'file-system'

export const iterateOverWorkspacesAndReadFiles = curry(
  function _iterateOverWorkspacesAndReadFiles(
    { search: searchGlob, ignore },
    root,
    x
  ) {
    log.reader('searchGlob?', searchGlob)
    return pipe(
      // look for specific file types
      map(workspace => workspace + searchGlob),
      log.core('ITERATING'),
      chain(
        readDirWithConfig({
          // exclude some search spaces
          ignore,
          cwd: root,
        })
      )
    )(x)
  }
)

export const readPackageJsonWorkspaces = curry(
  function _readPackageJsonWorkspaces(root, x) {
    return map(
      pipe(
        propOr([], 'workspaces'),
        log.parse('workspaces'),
        // we want directories only
        map(z => `${z}/`),
        map(readDirWithConfig({ cwd: root }))
      )
    )(x)
  }
)

export const monorepoRunner = curry(
  function _monorepoRunnerF(cancel, config, root, pkgJsonPath) {
    const { showMatchesOnly } = config
    log.core('CONFIG', config)
    return pipe(
      log.core('reading root package.json'),
      readJSONFile,
      readPackageJsonWorkspaces(root),
      chain(parallel(10)),
      signal(cancel, {
        text: 'Reading all workspaces...',
        successText: 'Read all workspaces!',
        failText: 'Unable to read all workspaces.',
      }),
      map(
        pipe(
          flatten,
          log.core('workflows, flat'),
          iterateOverWorkspacesAndReadFiles(config, root)
        )
      ),
      chain(parallel(10)),
      when(
        () => !showMatchesOnly,
        pipe(
          map(flatten),
          log.verbose('all files read'),
          signal(cancel, {
            text: 'Reading all files...',
            successText: 'Read all files!',
            failText: 'Unable to read all files.',
          })
        )
      )
    )(pkgJsonPath)
  }
)
