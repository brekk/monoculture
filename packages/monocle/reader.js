import path from 'node:path'

import {
  identity as I,
  curry,
  pipe,
  map,
  trim,
  split,
  addIndex,
  chain,
} from 'ramda'
import { parallel } from 'fluture'

import { readDirWithConfig, readFile } from 'file-system'
import { futureFileProcessor } from 'monorail'

import { hash } from './hash'
import { log } from './log'

export const readMonoFile = curry(
  function _readMonoFile(basePath, trimContent, file) {
    return pipe(
      log.file('reading'),
      readFile,
      map(buf =>
        pipe(
          split('\n'),
          // ostensibly files have content that starts at a 1 index
          addIndex(map)((y, i) => [i + 1, trimContent ? trim(y) : y]),
          body => ({
            name: path.relative(basePath, file),
            hash: hash(buf),
            body,
          })
        )(buf)
      )
    )(file)
  }
)

export const readAll = curry(function _readAll(config, dirglob) {
  return pipe(
    log.file('reading glob'),
    readDirWithConfig({ ...config, nodir: true }),
    config.showMatchesOnly
      ? I
      : chain(files =>
          pipe(map(readMonoFile(config.cwd, config.trim)), parallel(10))(files)
        )
  )(dirglob)
})

export const monoprocessor = curry(
  function _monoprocessor(config, pluginsF, dirGlob) {
    return pipe(
      readAll(config),
      config.showMatchesOnly ? I : futureFileProcessor(config, pluginsF)
    )(dirGlob)
  }
)
