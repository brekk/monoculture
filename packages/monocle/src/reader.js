import path from 'node:path'

import {
  identity as I,
  curry,
  pipe,
  map,
  trim,
  split,
  addIndex,
  chain,
} from 'ramda'
import { parallel } from 'fluture'

import { readDirWithConfig, readFile } from 'file-system'
import { futureFileProcessor } from 'monorail'

import { hash } from './hash'
import { log } from './trace'

export const readMonoFile = curry((basePath, trimContent, file) =>
  pipe(
    log.file('reading'),
    readFile,
    map(buf =>
      pipe(
        split('\n'),
        // ostensibly files have content that starts at a 1 index
        addIndex(map)((y, i) => [i + 1, trimContent ? trim(y) : y]),
        body => ({
          name: path.relative(basePath, file),
          hash: hash(buf),
          body,
        })
      )(buf)
    )
  )(file)
)

export const readAll = curry((config, dirglob) => {
  return pipe(
    log.file('reading glob'),
    readDirWithConfig({ ...config, nodir: true }),
    config.showMatchesOnly
      ? I
      : chain(files =>
          pipe(map(readMonoFile(config.cwd, config.trim)), parallel(10))(files)
        )
  )(dirglob)
})

export const monoprocessor = curry((config, pluginsF, dirGlob) =>
  pipe(
    readAll(config),
    config.showMatchesOnly ? I : futureFileProcessor(config, pluginsF)
  )(dirGlob)
)
