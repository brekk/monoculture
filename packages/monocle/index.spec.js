import path from 'node:path'
import { resolve, fork } from 'fluture'
import { pipe, map, prop, trim, reduce } from 'ramda'
import { readAll, monoprocessor } from './reader'
import * as EVERYTHING from './index'

test('default export', () => {
  expect(Object.keys(EVERYTHING)).toEqual([
    'hash',
    'monoprocessor',
    'readAll',
    'readMonoFile',
  ])
})
const DDD = process.cwd()
const IGNORE = [
  path.join(DDD, '*.spec.js'),
  '**/__snapshots__/**',
  '**/coverage/**',
  '**/dist/**',
]

test('readAll', done => {
  const files = readAll(
    {
      // if we don't ignore this file and snapshots,
      // we have an asymptotically infinite amount of effort
      ignore: IGNORE,
      basePath: DDD,
      cwd: DDD,
    },
    path.join(__dirname, '/*')
  )
  fork(done)(
    pipe(z => {
      const out = map(prop('hash'))(z)
      expect(out).toMatchSnapshot()
      done()
    })
  )(files)
})

const PLUGINS = [
  {
    default: {
      name: 'wordcount',
      dependencies: [],
      fn: (c, f) =>
        pipe(
          map(([, lineContent]) => lineContent.split(' ').map(trim).length),
          reduce((a, b) => a + b, 0)
        )(f.body),
    },
  },
  {
    default: {
      name: 'c-words',
      dependencies: [],
      preserveLine: true,
      fn: (c, line) => line.split(' ').filter(z => z.startsWith('c')).length,
    },
  },
]

test('monoprocessor', done => {
  fork(done)(x => {
    expect(x.state).toMatchSnapshot()
    done()
  })(
    monoprocessor(
      {
        basePath: __dirname,
        cwd: __dirname,
        nodir: true,
        // don't read test files, don't read snapshots
        ignore: IGNORE,
      },
      resolve(PLUGINS),
      path.join(__dirname, '/**')
    )
  )
})
