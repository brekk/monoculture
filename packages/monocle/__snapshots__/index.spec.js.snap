// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`monoprocessor 1`] = `
{
  "c-words": {
    "README.md": {
      "body": [
        [
          1,
          0,
        ],
        [
          2,
          0,
        ],
        [
          3,
          0,
        ],
        [
          4,
          0,
        ],
        [
          5,
          2,
        ],
        [
          6,
          0,
        ],
        [
          7,
          1,
        ],
        [
          8,
          0,
        ],
        [
          9,
          0,
        ],
        [
          10,
          0,
        ],
        [
          11,
          0,
        ],
        [
          12,
          0,
        ],
        [
          13,
          0,
        ],
        [
          14,
          0,
        ],
        [
          15,
          0,
        ],
        [
          16,
          1,
        ],
        [
          17,
          1,
        ],
        [
          18,
          0,
        ],
        [
          19,
          0,
        ],
        [
          20,
          0,
        ],
        [
          21,
          0,
        ],
        [
          22,
          0,
        ],
        [
          23,
          2,
        ],
        [
          24,
          0,
        ],
        [
          25,
          0,
        ],
      ],
      "hash": "cecf9cfa2850eecfc09b728424354d72e4250e5c1838ee2f7cb0f0daaa144b4d",
      "name": "README.md",
    },
    "cli.js": {
      "body": [
        [
          1,
          0,
        ],
        [
          2,
          0,
        ],
        [
          3,
          0,
        ],
        [
          4,
          0,
        ],
        [
          5,
          0,
        ],
        [
          6,
          0,
        ],
        [
          7,
          1,
        ],
        [
          8,
          0,
        ],
        [
          9,
          0,
        ],
        [
          10,
          1,
        ],
        [
          11,
          0,
        ],
        [
          12,
          0,
        ],
        [
          13,
          0,
        ],
        [
          14,
          0,
        ],
        [
          15,
          0,
        ],
        [
          16,
          0,
        ],
        [
          17,
          2,
        ],
        [
          18,
          0,
        ],
        [
          19,
          0,
        ],
        [
          20,
          0,
        ],
        [
          21,
          0,
        ],
        [
          22,
          0,
        ],
        [
          23,
          0,
        ],
        [
          24,
          1,
        ],
        [
          25,
          2,
        ],
        [
          26,
          0,
        ],
        [
          27,
          0,
        ],
        [
          28,
          3,
        ],
        [
          29,
          0,
        ],
        [
          30,
          1,
        ],
        [
          31,
          0,
        ],
        [
          32,
          1,
        ],
        [
          33,
          0,
        ],
        [
          34,
          0,
        ],
        [
          35,
          0,
        ],
        [
          36,
          1,
        ],
        [
          37,
          1,
        ],
        [
          38,
          2,
        ],
        [
          39,
          0,
        ],
        [
          40,
          1,
        ],
        [
          41,
          0,
        ],
        [
          42,
          0,
        ],
        [
          43,
          0,
        ],
        [
          44,
          0,
        ],
        [
          45,
          0,
        ],
        [
          46,
          0,
        ],
        [
          47,
          0,
        ],
        [
          48,
          0,
        ],
        [
          49,
          0,
        ],
        [
          50,
          0,
        ],
        [
          51,
          0,
        ],
        [
          52,
          0,
        ],
        [
          53,
          0,
        ],
        [
          54,
          0,
        ],
        [
          55,
          1,
        ],
        [
          56,
          0,
        ],
        [
          57,
          0,
        ],
        [
          58,
          0,
        ],
        [
          59,
          0,
        ],
        [
          60,
          0,
        ],
        [
          61,
          0,
        ],
        [
          62,
          1,
        ],
        [
          63,
          3,
        ],
        [
          64,
          2,
        ],
        [
          65,
          1,
        ],
        [
          66,
          1,
        ],
        [
          67,
          1,
        ],
        [
          68,
          0,
        ],
        [
          69,
          2,
        ],
        [
          70,
          1,
        ],
        [
          71,
          0,
        ],
        [
          72,
          1,
        ],
        [
          73,
          0,
        ],
        [
          74,
          0,
        ],
        [
          75,
          0,
        ],
        [
          76,
          0,
        ],
        [
          77,
          0,
        ],
        [
          78,
          0,
        ],
        [
          79,
          0,
        ],
        [
          80,
          1,
        ],
        [
          81,
          1,
        ],
        [
          82,
          0,
        ],
        [
          83,
          0,
        ],
        [
          84,
          0,
        ],
        [
          85,
          0,
        ],
        [
          86,
          0,
        ],
        [
          87,
          0,
        ],
        [
          88,
          0,
        ],
        [
          89,
          0,
        ],
        [
          90,
          0,
        ],
        [
          91,
          0,
        ],
        [
          92,
          1,
        ],
        [
          93,
          0,
        ],
        [
          94,
          0,
        ],
        [
          95,
          0,
        ],
        [
          96,
          1,
        ],
        [
          97,
          0,
        ],
        [
          98,
          0,
        ],
        [
          99,
          0,
        ],
        [
          100,
          0,
        ],
        [
          101,
          0,
        ],
        [
          102,
          0,
        ],
        [
          103,
          0,
        ],
        [
          104,
          0,
        ],
        [
          105,
          0,
        ],
        [
          106,
          0,
        ],
        [
          107,
          0,
        ],
        [
          108,
          0,
        ],
        [
          109,
          0,
        ],
        [
          110,
          1,
        ],
        [
          111,
          0,
        ],
        [
          112,
          0,
        ],
        [
          113,
          0,
        ],
      ],
      "hash": "d90d1a8245ef3c70484219ccb9c44c310d08a583bb02501e0a20d5e631b2b25a",
      "name": "cli.js",
    },
    "config.js": {
      "body": [
        [
          1,
          1,
        ],
        [
          2,
          0,
        ],
        [
          3,
          0,
        ],
        [
          4,
          0,
        ],
        [
          5,
          0,
        ],
        [
          6,
          0,
        ],
        [
          7,
          0,
        ],
        [
          8,
          0,
        ],
        [
          9,
          0,
        ],
        [
          10,
          0,
        ],
        [
          11,
          0,
        ],
        [
          12,
          0,
        ],
        [
          13,
          0,
        ],
        [
          14,
          1,
        ],
        [
          15,
          0,
        ],
        [
          16,
          0,
        ],
        [
          17,
          0,
        ],
        [
          18,
          0,
        ],
        [
          19,
          1,
        ],
        [
          20,
          0,
        ],
        [
          21,
          0,
        ],
        [
          22,
          0,
        ],
        [
          23,
          0,
        ],
        [
          24,
          1,
        ],
        [
          25,
          0,
        ],
        [
          26,
          0,
        ],
        [
          27,
          0,
        ],
        [
          28,
          0,
        ],
        [
          29,
          1,
        ],
        [
          30,
          0,
        ],
        [
          31,
          0,
        ],
        [
          32,
          0,
        ],
        [
          33,
          1,
        ],
        [
          34,
          0,
        ],
        [
          35,
          2,
        ],
        [
          36,
          0,
        ],
        [
          37,
          0,
        ],
        [
          38,
          0,
        ],
        [
          39,
          1,
        ],
        [
          40,
          0,
        ],
        [
          41,
          0,
        ],
        [
          42,
          1,
        ],
        [
          43,
          0,
        ],
        [
          44,
          0,
        ],
        [
          45,
          1,
        ],
        [
          46,
          0,
        ],
        [
          47,
          0,
        ],
        [
          48,
          0,
        ],
        [
          49,
          0,
        ],
        [
          50,
          0,
        ],
        [
          51,
          0,
        ],
        [
          52,
          0,
        ],
        [
          53,
          0,
        ],
      ],
      "hash": "8d8164b8400b0d41daf95e4bc7904de22a7b4fdef1440a00a0fe61e4640d5474",
      "name": "config.js",
    },
    "examples/plugin-c-words.js": {
      "body": [
        [
          1,
          1,
        ],
        [
          2,
          0,
        ],
        [
          3,
          0,
        ],
        [
          4,
          0,
        ],
        [
          5,
          0,
        ],
        [
          6,
          0,
        ],
        [
          7,
          0,
        ],
        [
          8,
          0,
        ],
        [
          9,
          0,
        ],
      ],
      "hash": "730f1b2b11f3c1e427e32aa8f2c375efdfb306d38caeca77c8fde2aea4ce63b9",
      "name": "examples/plugin-c-words.js",
    },
    "examples/plugin-get-exports.js": {
      "body": [
        [
          1,
          0,
        ],
        [
          2,
          1,
        ],
        [
          3,
          1,
        ],
        [
          4,
          1,
        ],
        [
          5,
          0,
        ],
        [
          6,
          0,
        ],
        [
          7,
          0,
        ],
        [
          8,
          0,
        ],
        [
          9,
          1,
        ],
        [
          10,
          1,
        ],
        [
          11,
          1,
        ],
        [
          12,
          1,
        ],
        [
          13,
          0,
        ],
        [
          14,
          0,
        ],
        [
          15,
          0,
        ],
        [
          16,
          0,
        ],
        [
          17,
          0,
        ],
        [
          18,
          0,
        ],
        [
          19,
          0,
        ],
        [
          20,
          0,
        ],
        [
          21,
          0,
        ],
        [
          22,
          0,
        ],
        [
          23,
          0,
        ],
        [
          24,
          0,
        ],
      ],
      "hash": "cd55fedb68dc8ac2ec4a1a53a0626e078d8773abefcc04ac3cc7af1115261474",
      "name": "examples/plugin-get-exports.js",
    },
    "examples/plugin-get-imports.js": {
      "body": [
        [
          1,
          1,
        ],
        [
          2,
          0,
        ],
        [
          3,
          0,
        ],
        [
          4,
          0,
        ],
        [
          5,
          1,
        ],
        [
          6,
          1,
        ],
        [
          7,
          0,
        ],
        [
          8,
          0,
        ],
        [
          9,
          0,
        ],
        [
          10,
          0,
        ],
        [
          11,
          0,
        ],
        [
          12,
          0,
        ],
        [
          13,
          0,
        ],
        [
          14,
          0,
        ],
        [
          15,
          0,
        ],
        [
          16,
          0,
        ],
      ],
      "hash": "1bf3d8b3f4401c874059caeb10bcabd1464d5c63da3e198159aefb93649b0d19",
      "name": "examples/plugin-get-imports.js",
    },
    "examples/plugin-has-exports.js": {
      "body": [
        [
          1,
          0,
        ],
        [
          2,
          1,
        ],
        [
          3,
          0,
        ],
        [
          4,
          0,
        ],
        [
          5,
          0,
        ],
        [
          6,
          0,
        ],
        [
          7,
          0,
        ],
        [
          8,
          0,
        ],
        [
          9,
          0,
        ],
      ],
      "hash": "79c2a23fda02f19ee74091354f558d4fb0709b5c10943b70e289d80e8a18b895",
      "name": "examples/plugin-has-exports.js",
    },
    "examples/plugin-has-imports.js": {
      "body": [
        [
          1,
          0,
        ],
        [
          2,
          1,
        ],
        [
          3,
          0,
        ],
        [
          4,
          0,
        ],
        [
          5,
          0,
        ],
        [
          6,
          0,
        ],
        [
          7,
          0,
        ],
        [
          8,
          0,
        ],
        [
          9,
          0,
        ],
        [
          10,
          0,
        ],
      ],
      "hash": "ca67db8861461853c73e40456eb4d81042631d330339550dfe54323045c14ad0",
      "name": "examples/plugin-has-imports.js",
    },
    "examples/plugin-wordcount.js": {
      "body": [
        [
          1,
          0,
        ],
        [
          2,
          0,
        ],
        [
          3,
          1,
        ],
        [
          4,
          0,
        ],
        [
          5,
          0,
        ],
        [
          6,
          0,
        ],
        [
          7,
          0,
        ],
        [
          8,
          0,
        ],
        [
          9,
          0,
        ],
        [
          10,
          0,
        ],
        [
          11,
          0,
        ],
        [
          12,
          0,
        ],
        [
          13,
          0,
        ],
        [
          14,
          0,
        ],
      ],
      "hash": "eb5f935581c920769b6ee893bf75d197ca58f7477eb5344a91b659a8ec1b3013",
      "name": "examples/plugin-wordcount.js",
    },
    "examples/rulefile-test-run.json": {
      "body": [
        [
          1,
          0,
        ],
        [
          2,
          0,
        ],
        [
          3,
          0,
        ],
        [
          4,
          0,
        ],
        [
          5,
          0,
        ],
        [
          6,
          0,
        ],
        [
          7,
          0,
        ],
        [
          8,
          0,
        ],
        [
          9,
          0,
        ],
        [
          10,
          0,
        ],
        [
          11,
          0,
        ],
        [
          12,
          0,
        ],
        [
          13,
          0,
        ],
        [
          14,
          0,
        ],
        [
          15,
          0,
        ],
        [
          16,
          0,
        ],
        [
          17,
          0,
        ],
        [
          18,
          0,
        ],
        [
          19,
          0,
        ],
        [
          20,
          0,
        ],
        [
          21,
          0,
        ],
        [
          22,
          0,
        ],
      ],
      "hash": "dfa3bd9a3947edd635423c456de5953e341f624e2a60bc48223db0992d8f621d",
      "name": "examples/rulefile-test-run.json",
    },
    "examples/rulefile-test-run.toml": {
      "body": [
        [
          1,
          0,
        ],
        [
          2,
          0,
        ],
        [
          3,
          0,
        ],
        [
          4,
          0,
        ],
        [
          5,
          0,
        ],
        [
          6,
          0,
        ],
        [
          7,
          0,
        ],
        [
          8,
          0,
        ],
        [
          9,
          0,
        ],
        [
          10,
          0,
        ],
        [
          11,
          0,
        ],
        [
          12,
          0,
        ],
        [
          13,
          0,
        ],
        [
          14,
          0,
        ],
        [
          15,
          0,
        ],
        [
          16,
          0,
        ],
        [
          17,
          0,
        ],
        [
          18,
          0,
        ],
        [
          19,
          0,
        ],
        [
          20,
          0,
        ],
        [
          21,
          0,
        ],
        [
          22,
          0,
        ],
        [
          23,
          0,
        ],
      ],
      "hash": "2fbe237f240c40874d45938de3b4488aeab3ad44be6036730a2a194e078d9e20",
      "name": "examples/rulefile-test-run.toml",
    },
    "graph.svg": {
      "body": [
        [
          1,
          0,
        ],
        [
          2,
          0,
        ],
        [
          3,
          0,
        ],
        [
          4,
          0,
        ],
        [
          5,
          0,
        ],
        [
          6,
          0,
        ],
        [
          7,
          0,
        ],
        [
          8,
          0,
        ],
        [
          9,
          1,
        ],
        [
          10,
          0,
        ],
        [
          11,
          0,
        ],
        [
          12,
          1,
        ],
        [
          13,
          1,
        ],
        [
          14,
          0,
        ],
        [
          15,
          0,
        ],
        [
          16,
          0,
        ],
        [
          17,
          0,
        ],
        [
          18,
          1,
        ],
        [
          19,
          1,
        ],
        [
          20,
          0,
        ],
        [
          21,
          0,
        ],
        [
          22,
          0,
        ],
        [
          23,
          0,
        ],
        [
          24,
          1,
        ],
        [
          25,
          1,
        ],
        [
          26,
          0,
        ],
        [
          27,
          0,
        ],
        [
          28,
          0,
        ],
        [
          29,
          0,
        ],
        [
          30,
          0,
        ],
        [
          31,
          1,
        ],
        [
          32,
          0,
        ],
        [
          33,
          0,
        ],
        [
          34,
          0,
        ],
        [
          35,
          0,
        ],
        [
          36,
          1,
        ],
        [
          37,
          1,
        ],
        [
          38,
          0,
        ],
        [
          39,
          0,
        ],
        [
          40,
          0,
        ],
        [
          41,
          0,
        ],
        [
          42,
          0,
        ],
        [
          43,
          1,
        ],
        [
          44,
          0,
        ],
        [
          45,
          0,
        ],
        [
          46,
          0,
        ],
        [
          47,
          0,
        ],
        [
          48,
          1,
        ],
        [
          49,
          1,
        ],
        [
          50,
          0,
        ],
        [
          51,
          0,
        ],
        [
          52,
          0,
        ],
        [
          53,
          0,
        ],
        [
          54,
          0,
        ],
        [
          55,
          1,
        ],
        [
          56,
          0,
        ],
        [
          57,
          0,
        ],
        [
          58,
          0,
        ],
        [
          59,
          0,
        ],
        [
          60,
          1,
        ],
        [
          61,
          1,
        ],
        [
          62,
          0,
        ],
        [
          63,
          0,
        ],
        [
          64,
          0,
        ],
        [
          65,
          0,
        ],
        [
          66,
          0,
        ],
        [
          67,
          1,
        ],
        [
          68,
          0,
        ],
        [
          69,
          0,
        ],
        [
          70,
          0,
        ],
        [
          71,
          0,
        ],
        [
          72,
          0,
        ],
        [
          73,
          1,
        ],
        [
          74,
          0,
        ],
        [
          75,
          0,
        ],
        [
          76,
          0,
        ],
        [
          77,
          0,
        ],
        [
          78,
          0,
        ],
        [
          79,
          1,
        ],
        [
          80,
          0,
        ],
        [
          81,
          0,
        ],
        [
          82,
          0,
        ],
        [
          83,
          0,
        ],
        [
          84,
          0,
        ],
        [
          85,
          0,
        ],
        [
          86,
          0,
        ],
      ],
      "hash": "fa3c73824fe4e8bbdedcb0b6349190cf87a061232f53c86e5ee1c3011606f469",
      "name": "graph.svg",
    },
    "hash.js": {
      "body": [
        [
          1,
          1,
        ],
        [
          2,
          0,
        ],
        [
          3,
          1,
        ],
        [
          4,
          2,
        ],
        [
          5,
          0,
        ],
        [
          6,
          0,
        ],
        [
          7,
          0,
        ],
        [
          8,
          0,
        ],
      ],
      "hash": "5cd99d9df3d8e4c2fcc83cf248314a4836c71d88f8d6e4fdc0d84c4e8c7b029d",
      "name": "hash.js",
    },
    "index.js": {
      "body": [
        [
          1,
          0,
        ],
        [
          2,
          0,
        ],
        [
          3,
          0,
        ],
      ],
      "hash": "f5ca9070a68f19933e3cd35a10e18013655b9b9a9c1b78ec2c7262a98c2fee17",
      "name": "index.js",
    },
    "jest.config.cjs": {
      "body": [
        [
          1,
          2,
        ],
        [
          2,
          0,
        ],
        [
          3,
          0,
        ],
        [
          4,
          0,
        ],
        [
          5,
          0,
        ],
      ],
      "hash": "af210769222c38f339d77fa8ac1638aead4c191818f39f1f5c754ada22a980be",
      "name": "jest.config.cjs",
    },
    "package-scripts.cjs": {
      "body": [
        [
          1,
          1,
        ],
        [
          2,
          0,
        ],
        [
          3,
          1,
        ],
        [
          4,
          0,
        ],
        [
          5,
          0,
        ],
        [
          6,
          0,
        ],
        [
          7,
          0,
        ],
        [
          8,
          0,
        ],
        [
          9,
          0,
        ],
        [
          10,
          0,
        ],
        [
          11,
          0,
        ],
        [
          12,
          0,
        ],
        [
          13,
          0,
        ],
        [
          14,
          0,
        ],
        [
          15,
          1,
        ],
        [
          16,
          1,
        ],
        [
          17,
          0,
        ],
        [
          18,
          1,
        ],
        [
          19,
          0,
        ],
        [
          20,
          0,
        ],
        [
          21,
          0,
        ],
        [
          22,
          0,
        ],
        [
          23,
          1,
        ],
        [
          24,
          0,
        ],
        [
          25,
          1,
        ],
        [
          26,
          0,
        ],
        [
          27,
          0,
        ],
        [
          28,
          0,
        ],
        [
          29,
          0,
        ],
        [
          30,
          0,
        ],
        [
          31,
          0,
        ],
        [
          32,
          0,
        ],
        [
          33,
          0,
        ],
        [
          34,
          0,
        ],
        [
          35,
          1,
        ],
        [
          36,
          0,
        ],
        [
          37,
          0,
        ],
        [
          38,
          0,
        ],
        [
          39,
          0,
        ],
        [
          40,
          0,
        ],
        [
          41,
          0,
        ],
        [
          42,
          0,
        ],
        [
          43,
          0,
        ],
        [
          44,
          0,
        ],
        [
          45,
          0,
        ],
        [
          46,
          0,
        ],
        [
          47,
          0,
        ],
        [
          48,
          0,
        ],
        [
          49,
          0,
        ],
        [
          50,
          0,
        ],
        [
          51,
          0,
        ],
      ],
      "hash": "cb508146748ce89a70fc1e66982441a82c714d514275c9b49b57fd79b00ca0b7",
      "name": "package-scripts.cjs",
    },
    "package.json": {
      "body": [
        [
          1,
          0,
        ],
        [
          2,
          0,
        ],
        [
          3,
          0,
        ],
        [
          4,
          1,
        ],
        [
          5,
          0,
        ],
        [
          6,
          0,
        ],
        [
          7,
          0,
        ],
        [
          8,
          0,
        ],
        [
          9,
          0,
        ],
        [
          10,
          0,
        ],
        [
          11,
          0,
        ],
        [
          12,
          0,
        ],
        [
          13,
          0,
        ],
        [
          14,
          0,
        ],
        [
          15,
          0,
        ],
        [
          16,
          0,
        ],
        [
          17,
          0,
        ],
        [
          18,
          0,
        ],
        [
          19,
          0,
        ],
        [
          20,
          0,
        ],
        [
          21,
          0,
        ],
        [
          22,
          0,
        ],
        [
          23,
          0,
        ],
        [
          24,
          0,
        ],
        [
          25,
          0,
        ],
        [
          26,
          0,
        ],
        [
          27,
          0,
        ],
        [
          28,
          0,
        ],
        [
          29,
          0,
        ],
        [
          30,
          0,
        ],
        [
          31,
          0,
        ],
        [
          32,
          0,
        ],
        [
          33,
          0,
        ],
        [
          34,
          0,
        ],
        [
          35,
          0,
        ],
        [
          36,
          1,
        ],
        [
          37,
          0,
        ],
        [
          38,
          0,
        ],
        [
          39,
          0,
        ],
        [
          40,
          0,
        ],
        [
          41,
          0,
        ],
        [
          42,
          0,
        ],
        [
          43,
          0,
        ],
        [
          44,
          0,
        ],
        [
          45,
          0,
        ],
        [
          46,
          0,
        ],
        [
          47,
          0,
        ],
      ],
      "hash": "f58c58469a7aafdaf453c4caf867fb922747bf8f2cc8071869715e7d6b86a1b0",
      "name": "package.json",
    },
    "reader.js": {
      "body": [
        [
          1,
          0,
        ],
        [
          2,
          0,
        ],
        [
          3,
          0,
        ],
        [
          4,
          0,
        ],
        [
          5,
          1,
        ],
        [
          6,
          0,
        ],
        [
          7,
          0,
        ],
        [
          8,
          0,
        ],
        [
          9,
          0,
        ],
        [
          10,
          0,
        ],
        [
          11,
          1,
        ],
        [
          12,
          0,
        ],
        [
          13,
          0,
        ],
        [
          14,
          0,
        ],
        [
          15,
          0,
        ],
        [
          16,
          0,
        ],
        [
          17,
          0,
        ],
        [
          18,
          0,
        ],
        [
          19,
          0,
        ],
        [
          20,
          0,
        ],
        [
          21,
          2,
        ],
        [
          22,
          0,
        ],
        [
          23,
          0,
        ],
        [
          24,
          0,
        ],
        [
          25,
          0,
        ],
        [
          26,
          0,
        ],
        [
          27,
          0,
        ],
        [
          28,
          1,
        ],
        [
          29,
          0,
        ],
        [
          30,
          0,
        ],
        [
          31,
          0,
        ],
        [
          32,
          0,
        ],
        [
          33,
          0,
        ],
        [
          34,
          0,
        ],
        [
          35,
          0,
        ],
        [
          36,
          0,
        ],
        [
          37,
          0,
        ],
        [
          38,
          0,
        ],
        [
          39,
          0,
        ],
        [
          40,
          2,
        ],
        [
          41,
          0,
        ],
        [
          42,
          0,
        ],
        [
          43,
          0,
        ],
        [
          44,
          1,
        ],
        [
          45,
          0,
        ],
        [
          46,
          1,
        ],
        [
          47,
          1,
        ],
        [
          48,
          0,
        ],
        [
          49,
          0,
        ],
        [
          50,
          0,
        ],
        [
          51,
          0,
        ],
        [
          52,
          2,
        ],
        [
          53,
          0,
        ],
        [
          54,
          0,
        ],
        [
          55,
          1,
        ],
        [
          56,
          0,
        ],
        [
          57,
          0,
        ],
        [
          58,
          0,
        ],
      ],
      "hash": "d407873d6c8d31716c263d133720e5f34356dbfafdcbdaed734f6ee341791286",
      "name": "reader.js",
    },
    "trace.js": {
      "body": [
        [
          1,
          1,
        ],
        [
          2,
          0,
        ],
        [
          3,
          2,
        ],
        [
          4,
          0,
        ],
      ],
      "hash": "97c92175e3d025641201b1fdb1fbb7c7852f980f1b73890b4b8e07a7184ef2ba",
      "name": "trace.js",
    },
  },
  "wordcount": {
    "README.md": 210,
    "cli.js": 1068,
    "config.js": 351,
    "examples/plugin-c-words.js": 32,
    "examples/plugin-get-exports.js": 222,
    "examples/plugin-get-imports.js": 96,
    "examples/plugin-has-exports.js": 36,
    "examples/plugin-has-imports.js": 42,
    "examples/plugin-wordcount.js": 68,
    "examples/rulefile-test-run.json": 90,
    "examples/rulefile-test-run.toml": 69,
    "graph.svg": 438,
    "hash.js": 27,
    "index.js": 9,
    "jest.config.cjs": 12,
    "package-scripts.cjs": 345,
    "package.json": 289,
    "reader.js": 331,
    "trace.js": 16,
  },
}
`;

exports[`readAll 1`] = `
[
  "d90d1a8245ef3c70484219ccb9c44c310d08a583bb02501e0a20d5e631b2b25a",
  "8d8164b8400b0d41daf95e4bc7904de22a7b4fdef1440a00a0fe61e4640d5474",
  "fa3c73824fe4e8bbdedcb0b6349190cf87a061232f53c86e5ee1c3011606f469",
  "5cd99d9df3d8e4c2fcc83cf248314a4836c71d88f8d6e4fdc0d84c4e8c7b029d",
  "f5ca9070a68f19933e3cd35a10e18013655b9b9a9c1b78ec2c7262a98c2fee17",
  "af210769222c38f339d77fa8ac1638aead4c191818f39f1f5c754ada22a980be",
  "cb508146748ce89a70fc1e66982441a82c714d514275c9b49b57fd79b00ca0b7",
  "f58c58469a7aafdaf453c4caf867fb922747bf8f2cc8071869715e7d6b86a1b0",
  "d407873d6c8d31716c263d133720e5f34356dbfafdcbdaed734f6ee341791286",
  "cecf9cfa2850eecfc09b728424354d72e4250e5c1838ee2f7cb0f0daaa144b4d",
  "97c92175e3d025641201b1fdb1fbb7c7852f980f1b73890b4b8e07a7184ef2ba",
]
`;
